tests/unit/test_library_smtp.py

```diff
@@ -0,0 +1,272 @@
+        "host": "example.smtp",
+        "port": "25",
+        "user": "example_user",
+        "password": "somepassword",  # nosec
+        "auth_type": "plain",
+        "transport_security": "tls",
+        "domain": "domain",
+    }
+
+    harness = Harness(SmtpRequirerCharm, meta=REQUIRER_METADATA)
+    harness.begin()
+    harness.set_leader(is_leader)
+    harness.add_relation("smtp-legacy", "smtp-provider", app_data=relation_data)
+
+    assert len(harness.charm.events) == 1
+    assert harness.charm.events[0].host == relation_data["host"]
+    assert harness.charm.events[0].port == int(relation_data["port"])
+    assert harness.charm.events[0].user == relation_data["user"]
+    assert harness.charm.events[0].password == relation_data["password"]
+    assert harness.charm.events[0].auth_type == relation_data["auth_type"]
+    assert harness.charm.events[0].transport_security == relation_data["transport_security"]
+    assert harness.charm.events[0].domain == relation_data["domain"]
+
+    retrieved_relation_data = harness.charm.smtp_legacy.get_relation_data()
+    assert retrieved_relation_data.host == relation_data["host"]
+    assert retrieved_relation_data.port == int(relation_data["port"])
+    assert retrieved_relation_data.user == relation_data["user"]
+    assert retrieved_relation_data.password == relation_data["password"]
+    assert retrieved_relation_data.auth_type == relation_data["auth_type"]
+    assert retrieved_relation_data.transport_security == relation_data["transport_security"]
+    assert retrieved_relation_data.domain == relation_data["domain"]
+
+
+@pytest.mark.parametrize("is_leader", [True, False])
+def test_requirer_charm_with_valid_relation_data_emits_event(is_leader):
+    """
+    arrange: set up a charm.
+    act: trigger a relation changed event with valid data.
+    assert: a event containing the relation data is emitted.
+    """
+    relation_data = {
+        "host": "example.smtp",
+        "port": "25",
+        "user": "example_user",
+        "password_id": "someid",
+        "auth_type": "plain",
+        "transport_security": "tls",
+        "domain": "domain",
+    }
+
+    harness = Harness(SmtpRequirerCharm, meta=REQUIRER_METADATA)
+    harness.begin()
+    harness.set_leader(is_leader)
+    harness.add_relation("smtp", "smtp-provider", app_data=relation_data)
+
+    assert len(harness.charm.events) == 1
+    assert harness.charm.events[0].host == relation_data["host"]
+    assert harness.charm.events[0].port == int(relation_data["port"])
+    assert harness.charm.events[0].user == relation_data["user"]
+    assert harness.charm.events[0].password_id == relation_data["password_id"]
+    assert harness.charm.events[0].auth_type == relation_data["auth_type"]
+    assert harness.charm.events[0].transport_security == relation_data["transport_security"]
+    assert harness.charm.events[0].domain == relation_data["domain"]
+
+
+@pytest.mark.parametrize("is_leader", [True, False])
```

Comment:
As above.

Reply:
This is actually a regression test

Reply:
Same question as above. I don't see any relation between regression and the leadership.